src/Marten/Schema/SchemaBuilder.cs

```diff
@@ -37,15 +37,34 @@ public static void WriteSchemaObjects(DocumentMapping mapping, IDocumentSchema s
                 updates += ", " + duplicates.Select(x => $"{x.Column} = {x.Arg}").Join(", ");
             }
 
-
-            writer.WriteLine($"CREATE OR REPLACE FUNCTION {mapping.UpsertName}({argList}) RETURNS VOID AS");
-            writer.WriteLine("$$");
-            writer.WriteLine("BEGIN");
-            writer.WriteLine($"INSERT INTO {mapping.TableName} VALUES ({valueList})");
-            writer.WriteLine($"  ON CONFLICT ON CONSTRAINT pk_{mapping.TableName}");
-            writer.WriteLine($"  DO UPDATE SET {updates};");
-            writer.WriteLine("END;");
-            writer.WriteLine("$$ LANGUAGE plpgsql;");
+            if (schema != null && schema.UsePostgres9_4Fallback)
+            {
+                var inserts = "id, data";
```

Comment:
inserts are managed separately to address the requirements of the 9.4 fallback function